Remotion/Data/DomainObjects.UnitTests/Mapping/MappingConfigurationExtensionsTest.cs

```diff
@@ -0,0 +1,170 @@
+    public void GetClassDefinition ()
+    {
+      var classDefinition = ClassDefinitionObjectMother.CreateClassDefinition();
+      var mappingConfigurationMock = new Mock<IMappingConfiguration>(MockBehavior.Strict);
+      mappingConfigurationMock
+          .Setup(e => e.GetTypeDefinition(typeof(DummyClass), It.IsAny<Func<Type, Exception>>()))
+          .Returns(classDefinition);
+
+      var result = mappingConfigurationMock.Object.GetClassDefinition(typeof(DummyClass));
+      Assert.That(result, Is.SameAs(classDefinition));
+
+      mappingConfigurationMock.VerifyAll();
+    }
+
+    [Test]
+    public void GetClassDefinition_WithNonClass_Throws ()
+    {
+      var mappingConfigurationMock = new Mock<IMappingConfiguration>(MockBehavior.Strict);
+
+      Assert.That(
+          () => mappingConfigurationMock.Object.GetClassDefinition(typeof(IDummyInterface)),
+          Throws.TypeOf<ArgumentException>()
+              .With.ArgumentExceptionMessageEqualTo(
+                  "The specified type 'Remotion.Data.DomainObjects.UnitTests.Mapping.MappingConfigurationExtensionsTest+IDummyInterface' must be a class.",
+                  "type"));
+    }
+
+    [Test]
+    public void GetClassDefinition_WithCustomExceptionFactory ()
+    {
+      var typeDefinition = TypeDefinitionObjectMother.CreateClassDefinition();
+      var exceptionFactoryMock = new Mock<Func<Type, Exception>>(MockBehavior.Strict);
+      var mappingConfigurationMock = new Mock<IMappingConfiguration>(MockBehavior.Strict);
+      mappingConfigurationMock
+          .Setup(e => e.GetTypeDefinition(typeof(DummyClass), It.IsAny<Func<Type, Exception>>()))
+          .Returns(typeDefinition);
+
+      var result = mappingConfigurationMock.Object.GetClassDefinition(typeof(DummyClass), exceptionFactoryMock.Object);
+      Assert.That(result, Is.SameAs(typeDefinition));
+
+      mappingConfigurationMock.VerifyAll();
+    }
+
+    [Test]
+    public void GetClassDefinition_WithCustomerExceptionFactoryAndNonClass_Throws ()
+    {
+      var exceptionFactoryMock = new Mock<Func<Type, Exception>>(MockBehavior.Strict);
+      var mappingConfigurationMock = new Mock<IMappingConfiguration>(MockBehavior.Strict);
+
+      Assert.That(
+          () => mappingConfigurationMock.Object.GetClassDefinition(typeof(IDummyInterface), exceptionFactoryMock.Object),
+          Throws.TypeOf<ArgumentException>()
+              .With.ArgumentExceptionMessageEqualTo(
+                  "The specified type 'Remotion.Data.DomainObjects.UnitTests.Mapping.MappingConfigurationExtensionsTest+IDummyInterface' must be a class.",
+                  "type"));
+    }
+
+    [Test]
+    public void GetClassDefinition_WithCustomerExceptionFactoryAndNonExistingClass_ThrowsExceptionFromFactory ()
+    {
+      var exception = new Exception("Test");
+      var exceptionFactoryMock = new Mock<Func<Type, Exception>>(MockBehavior.Strict);
+      exceptionFactoryMock.Setup(e => e(typeof(DummyClass))).Returns(exception);
```

Comment:
But if we throw the exception from our own delegate the test won't fail if the SUT calls the delegate but does not throw the exception.

Reply:
> But if we throw the exception from our own delegate the test won't fail if the SUT calls the delegate but does not throw the exception.
That I don't get. We expect that the exception is thrown with Assert.That(...Throws.SameAs(expcetion). So, we have to call the delegate and use the result for the test to work. What am I missing?

Reply:
I dug into the styleguide and the topic of stub vs mock did already get a treaty last December and Verify vs VerifyAll did factor into how to migrated from RhinoMocks to Moq. We should incorporate this existing knowledge into the discussion to prevent going in circles.